lsp-protocol.el

```diff
@@ -0,0 +1,419 @@
+  failure-reason?) "
+  (->> interfaces
+       (-partition 3)
+       (-map (-lambda ((interface required optional))
+               (let ((params (nconc
+                              (-map (lambda (param-name)
+                                      (cons
+                                       (intern (concat ":" (s-dashed-words (symbol-name param-name)) "?"))
+                                       param-name))
+                                    optional)
+                              (-map (lambda (param-name)
+                                      (cons (intern (concat ":" (s-dashed-words (symbol-name param-name))))
+                                            param-name))
+                                    required))))
+                 (cl-list*
+                  `(defun ,(intern (format "dash-expand:&%s" interface)) (key source)
+                     (unless (member key ',(-map #'cl-first params))
+                       (error "Unknown key: %s. Available keys: %s" key ',(-map #'cl-first params)))
+                     `(gethash ,(substring (symbol-name
+                                            (cl-rest (assoc key ',params)))
+                                           1)
+                               ,source))
+                  `(defun ,(intern (format "lsp:%s?" (s-dashed-words (symbol-name interface)))) (object)
+                     (if (ht? object)
+                         (-all? (lambda (prop)
+                                  (ht-get object prop))
+                                ',(-map (lambda (field-name)
+                                          (substring (symbol-name field-name) 1))
+                                        required))
+                       (-all? (lambda (prop)
+                                (plist-get object prop))
+                              ',required)))
+                  (-mapcat (-lambda ((label . name))
+                             (list
+                              `(defun ,(intern (format "lsp:%s-%s"
+                                                       (s-dashed-words (symbol-name interface))
+                                                       (substring (symbol-name label) 1)))
+                                   (object)
+                                 (gethash ,(substring (symbol-name name) 1) object))
+                              `(defun ,(intern (format "lsp:set-%s-%s"
+                                                       (s-dashed-words (symbol-name interface))
+                                                       (substring (symbol-name label) 1)))
+                                   (object value)
+                                 (puthash ,(substring (symbol-name name) 1) value object))))
+                           params)))))
+       (apply #'append )
+       (list* 'progn)))
+
+
+
+(defconst lsp/completion-item-kind-text 0)
+;; begin autogenerated code
+
+(defconst lsp/completion-item-kind-text 0)
```

Comment:
Can we use `--` instead of `/`, are you afraid of name collision?

Reply:
It is either `-` or something else since it is public. I used `:` and `/` to make the code a bit more readable because you will have a visual clue.

Reply:
Thanks